### PR TITLE
Animations: block animation until all resources are loaded

### PR DESCRIPTION
**Pull request by dvoytenko**
/cc @emarchiori, @hellyhansen

### CODE REVIEW

File: extensions/amp-animation/0.1/web-animations.js [hunk 1]
```diff
@@ -169,9 +169,32 @@ export class MeasureScanner extends Scanner {
...
+   * @private
+   */
+  unblockElements_(resources) {
+    const promises = [];
+    for (let i = 0; i < this.targets_.length; i++) {
+      const element = this.targets_[i];
+      if (element.classList.contains('-amp-element')) {
+        const resource = resources.getResourceForElement(element);
+        // Ensure that the promise doesn't fail.
+        promises.push(resource.loadedOnce().catch(() => {}));
```
Comment (jridgewell) [inline extensions/amp-animation/0.1/web-animations.js]: `#loadedOnce` can never reject.

File: extensions/amp-animation/0.1/amp-animation.js [hunk 2]
```diff
@@ -66,14 +66,13 @@ export class AmpAnimation extends AMP.BaseElement {
...
-        this.runner_ = measurer.createRunner();
-        this.runner_.play();
-      },
+    const vsync = this.getVsync();
+    vsync.measurePromise(() => {
+      measurer.scan(configJson);
+      return measurer.createRunner(this.element.getResources());
+    }).then(runner => {
+      this.runner_ = runner;
+      this.runner_.play();
```
Comment (jridgewell) [inline extensions/amp-animation/0.1/amp-animation.js]: Does this need to be in a mutate block?

File: extensions/amp-animation/0.1/test/test-web-animations.js [hunk 3]
```diff
@@ -292,6 +292,65 @@ describes.sandboxed('MeasureScanner', {}, () => {
...
+      ]).then(res => {
+        runner = res;
+      });
+      return Promise.race([runnerPromise, Promise.resolve()]).then(() => {
+        expect(runner).to.be.ok;
+        expect(runner.requests_).to.have.length(2);
+      });
+    });
+
+    it('should block non-AMP elements immediately', () => {
```
Comment (aghassemi) [inline extensions/amp-animation/0.1/test/test-web-animations.js]: block AMP elements?

File: extensions/amp-animation/0.1/test/test-web-animations.js [hunk 4]
```diff
@@ -292,6 +292,65 @@ describes.sandboxed('MeasureScanner', {}, () => {
...
+
+    it('should unblock non-AMP elements immediately', () => {
+      let runner;
+      const runnerPromise = createRunner([
+        {target: target1, keyframes: {}},
+        {target: target2, keyframes: {}},
+      ]).then(res => {
+        runner = res;
+      });
+      return Promise.race([runnerPromise, Promise.resolve()]).then(() => {
```
Comment (aghassemi) [inline extensions/amp-animation/0.1/test/test-web-animations.js]: Why race? Promise.resolve().then( () => { //verify }); should technically be the same as the code above.

File: extensions/amp-animation/0.1/test/test-web-animations.js [hunk 5]
```diff
@@ -292,6 +292,65 @@ describes.sandboxed('MeasureScanner', {}, () => {
...
+
+    it('should block non-AMP elements immediately', () => {
+      let runner;
+      const runnerPromise = createRunner([
+        {target: amp1, keyframes: {}},
+        {target: amp2, keyframes: {}},
+      ]).then(res => {
+        runner = res;
+      });
+      return Promise.race([runnerPromise, Promise.resolve()]).then(() => {
```
Comment (aghassemi) [inline extensions/amp-animation/0.1/test/test-web-animations.js]: ditto unnecessary race.

File: extensions/amp-animation/0.1/web-animations.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-animation/0.1/web-animations.js]: Done

File: extensions/amp-animation/0.1/test/test-web-animations.js (hunk 4 above)
Comment (dvoytenko) [inline extensions/amp-animation/0.1/test/test-web-animations.js]: No technical reason. I just thought it communicates what I was waiting for better - i.e. `runnerPromise` is not resolved in that case.

File: extensions/amp-animation/0.1/test/test-web-animations.js (hunk 3 above)
Comment (dvoytenko) [inline extensions/amp-animation/0.1/test/test-web-animations.js]: Done

File: extensions/amp-animation/0.1/test/test-web-animations.js (hunk 5 above)
Comment (dvoytenko) [inline extensions/amp-animation/0.1/test/test-web-animations.js]: Done.

File: extensions/amp-animation/0.1/web-animations.js [hunk 6]
```diff
@@ -191,7 +191,7 @@ export class MeasureScanner extends Scanner {
       if (element.classList.contains('-amp-element')) {
         const resource = resources.getResourceForElement(element);
         // Ensure that the promise doesn't fail.
```
Comment (jridgewell) [inline extensions/amp-animation/0.1/web-animations.js]: Old comment.

Comment (dvoytenko) [inline extensions/amp-animation/0.1/web-animations.js]: Done

**Pull request merged**